main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala

```diff
@@ -342,8 +340,15 @@ object Simplifier {
       case TypedAst.Pattern.Int64(lit, loc) => SimplifiedAst.Expression.Int64(lit, loc)
       case TypedAst.Pattern.BigInt(lit, loc) => SimplifiedAst.Expression.BigInt(lit, loc)
       case TypedAst.Pattern.Str(lit, loc) => SimplifiedAst.Expression.Str(lit, loc)
-      case TypedAst.Pattern.Tag(sym, tag, p, tpe, loc) => SimplifiedAst.Expression.Tag(sym, tag, pat2exp(p), tpe, loc)
-      case TypedAst.Pattern.Tuple(elms, tpe, loc) => SimplifiedAst.Expression.Tuple(elms map pat2exp, tpe, loc)
+      case TypedAst.Pattern.Tag(sym, tag, p, tpe, loc) =>
+        val e = pat2exp(p)
+        SimplifiedAst.Expression.Tag(sym, tag, e, tpe, e.purity, loc)
+      case TypedAst.Pattern.Tuple(elms, tpe, loc) =>
+        val es = elms.map(pat2exp)
+        val purity = es.foldLeft[Purity](Pure){
```

Comment:
Introduce a function, after `combine`, called `combineAll` which takes a list of Purities and combines them.